src/picongpu/include/plugins/hdf5/WriteSpecies.hpp

```diff
@@ -157,48 +164,176 @@ struct WriteSpecies
+         * global rank or scalar position (which are not idential) are
+         * ordered as long as the particle attributes are also written in
+         * the same order (which is by global rank) */
+        uint64_t numParticlesOffset = 0;
+        uint64_t numParticlesGlobal = 0;
+
+        MPI_CHECK(MPI_Allgather(
+            myParticlePatch, 2, MPI_UINT64_T,
+            &(*particleCounts.begin()), 2, MPI_UINT64_T,
+            gc.getCommunicator().getMPIComm()
+        ));
+
+        for( uint64_t r = 0; r < gc.getGlobalSize(); ++r )
         {
-            ColTypeUInt64_5Array ctUInt64_5;
-            GridController<simDim>& gc = Environment<simDim>::get().GridController();
+            numParticlesGlobal += particleCounts.at(2 * r);
+            if( particleCounts.at(2 * r + 1) < myParticlePatch[ 1 ] )
+                numParticlesOffset += particleCounts.at(2 * r);
+        }
+        log<picLog::INPUT_OUTPUT > ("HDF5:  (end) collect particle sizes for %1%") % Hdf5FrameType::getName();
 
-            const size_t pos_offset = 2;
+        /* dump main particle data to hdf5 file */
+        log<picLog::INPUT_OUTPUT > ("HDF5:  (begin) write particle attributes for %1%") % Hdf5FrameType::getName();
 
-            /* particlesMetaInfo = (num particles, scalar position, particle offset x, y, z) */
-            uint64_t particlesMetaInfo[5] = {totalNumParticles, gc.getScalarPosition(), 0, 0, 0};
-            for (size_t d = 0; d < simDim; ++d)
-                particlesMetaInfo[pos_offset + d] = particleOffset[d];
+        ForEach<typename Hdf5FrameType::ValueTypeSeq, hdf5::ParticleAttribute<bmpl::_1> > writeToHdf5;
+        writeToHdf5(
+            params,
+            forward(hostFrame),
+            std::string("particles/") + FrameType::getName() + std::string("/") + subGroup,
+            numParticles,
+            numParticlesOffset,
+            numParticlesGlobal
+        );
 
-            /* prevent that top (y) gpus have negative value here */
-            if (gc.getPosition().y() == 0)
-                particlesMetaInfo[pos_offset + 1] = 0;
+        /* write meta attributes for species */
+        writeMetaAttributes(params);
 
-            if (particleOffset[1] < 0) // 1 == y
-                particlesMetaInfo[pos_offset + 1] = 0;
+        log<picLog::INPUT_OUTPUT > ("HDF5:  (end) write particle attributes for %1%") % Hdf5FrameType::getName();
+
+        /* write species particle patch meta information */
+        log<picLog::INPUT_OUTPUT > ("HDF5:  (begin) writing particlePatches for %1%") % Hdf5FrameType::getName();
+
+        std::string particlePatchesPath( std::string("particles/") +
+            FrameType::getName() + std::string("/") + subGroup +
+            std::string("/particlePatches") );
+
+        /* numParticles: number of particles in this patch */
+        params->dataCollector->write(
+            params->currentStep,
+            Dimensions(gc.getGlobalSize(), 1, 1),
```

Comment:
only a note: 
Because of the multiple usages of `gc.getGlobalSize()` and `gc.getGlobalRank()` a helper variable would help to read the code.
`gc.getGlobalRank()` == the offset inside the array which is written
`gc.getGlobalRank()` == the size of the array